core/node/external_proof_integration_api/src/metrics.rs

```diff
@@ -25,30 +27,37 @@ pub(crate) struct ProofIntegrationApiMetrics {
     pub call_latency: LabeledFamily<(Method, CallOutcome), Histogram<Duration>, 2>,
 }
 
-pub(crate) struct MethodCallGuard {
-    method_type: Method,
-    outcome: CallOutcome,
-    started_at: Instant,
-}
+pub(crate) async fn call_outcome_tracker(request: Request, next: Next) -> Response {
+    let start = Instant::now();
+    let path = request.uri().path();
 
-impl MethodCallGuard {
-    pub(crate) fn new(method_type: Method) -> Self {
-        MethodCallGuard {
-            method_type,
-            outcome: CallOutcome::Failure,
-            started_at: Instant::now(),
+    let method = if path.starts_with("/proof_generation_data") {
```

Comment:
You can improve the ergonomics by creating a factory for middlewares, e.g.:

```rust
    let middleware_factory = |method: Method| {
        axum::middleware::from_fn(move |req: Request, next: Next| async move {
            metrics_middleware(method);
            next.run(req).await
        })
    };

    let app = axum::Router::new()
        .route(
            "/handler_a",
            axum::routing::get(handler_a).layer(middleware_factory(Method::A)),
        )
        .route(
            "/handler_b/:input",
            axum::routing::get(handler_b).layer(middleware_factory(Method::B)),
        );
```

(note that it _has_ to be a closure, since it works with unnameable types)